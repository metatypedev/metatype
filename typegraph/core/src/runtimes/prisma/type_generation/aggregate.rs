// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::{
    errors::Result,
    t::{self, ConcreteTypeBuilder, TypeBuilder},
    types::{Type, TypeId},
};

use super::{TypeGen, TypeGenContext};

pub struct CountOutput {
    model_id: TypeId,
}

impl CountOutput {
    pub fn new(model_id: TypeId) -> Self {
        Self { model_id }
    }
}

impl TypeGen for CountOutput {
    fn generate(&self, _context: &mut TypeGenContext) -> Result<TypeId> {
        let mut builder = t::struct_();
        let opt_int = t::optional(t::integer().build()?).build()?;
        builder.prop("_all", opt_int);

        for (k, _) in self.model_id.as_struct()?.iter_props() {
            builder.prop(k, opt_int);
        }

        // TODO union
        builder.named(self.name()).build()
    }

    fn name(&self) -> String {
        let model_name = self.model_id.type_name().unwrap().unwrap();
        format!("_{}_AggrCount", model_name)
    }
}

pub struct NumberAggregateOutput {
    model_id: TypeId,
    avg: bool,
}

impl NumberAggregateOutput {
    pub fn new(model_id: TypeId, avg: bool) -> Self {
        Self { model_id, avg }
    }
}

impl TypeGen for NumberAggregateOutput {
    fn generate(&self, _context: &mut TypeGenContext) -> Result<TypeId> {
        let mut builder = t::struct_();

        let opt_float = t::optional(t::float().build()?).build()?;
        let for_int = if self.avg {
            opt_float
        } else {
            t::optional(t::integer().build()?).build()?
        };

        for (k, type_id) in self.model_id.as_struct()?.iter_props() {
            let type_id = type_id.non_optional_concrete_type()?;
            match type_id.as_type()? {
                Type::Integer(_) => {
                    builder.prop(k, for_int);
                }
                Type::Float(_) => {
                    builder.prop(k, opt_float);
                }
                Type::Optional(_) => unreachable!(),
                _ => {}
            }
        }

        builder.named(self.name()).build()
    }

    fn name(&self) -> String {
        let model_name = self.model_id.type_name().unwrap().unwrap();
        let suffix = if self.avg { "_avg" } else { "" };
        format!("_{model_name}_NumberAgg{suffix}")
    }
}
