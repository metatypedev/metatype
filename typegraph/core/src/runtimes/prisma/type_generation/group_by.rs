// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::{
    errors::Result,
    runtimes::prisma::type_generation::{where_::Where, with_filters::WithFilters},
    t::{self, ConcreteTypeBuilder, TypeBuilder},
    types::{Type, TypeId},
};

use super::{aggregate::CountOutput, TypeGen, TypeGenContext};

pub struct GroupingFields {
    model_id: TypeId,
}

impl GroupingFields {
    pub fn new(model_id: TypeId) -> Self {
        Self { model_id }
    }
}

impl TypeGen for GroupingFields {
    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId> {
        let mut fields = vec![];
        for (k, type_id) in self.model_id.as_struct()?.iter_props() {
            if context
                .registry
                .find_relationship_on(self.model_id, k)
                .is_some()
            {
                continue;
            }

            match type_id.non_optional_concrete_type()?.as_type()? {
                Type::Boolean(_)
                | Type::Integer(_)
                | Type::Float(_)
                | Type::String(_)
                | Type::Array(_) => fields.push(k.to_string()),
                _ => {}
            }
        }

        t::arrayx(t::string().enum_(fields))?
            .named(self.name())
            .build()
    }

    fn name(&self) -> String {
        let model_name = self.model_id.type_name().unwrap().unwrap();
        format!("_{}_GroupingFields", model_name)
    }
}

pub struct Having {
    model_id: TypeId,
}

impl Having {
    pub fn new(model_id: TypeId) -> Self {
        Self { model_id }
    }
}

impl TypeGen for Having {
    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId> {
        // TODO relations??
        let where_type = context.generate(&Where::new(self.model_id, false))?;
        let extended_type = context
            .generate(&WithFilters::new(where_type, self.model_id, true).with_aggregates())?;

        let name = self.name();
        let self_ref = t::proxy(&name).build()?;

        t::unionx![
            extended_type,
            t::struct_().propx("AND", t::array(self_ref))?,
            t::struct_().propx("OR", t::array(self_ref))?,
            t::struct_().prop("NOT", self_ref)
        ]
        .named(name)
        .build()
    }

    fn name(&self) -> String {
        let model_name = self.model_id.type_name().unwrap().unwrap();
        format!("_{}_Having", model_name)
    }
}

pub struct GroupByResult {
    model_id: TypeId,
}

impl GroupByResult {
    pub fn new(model_id: TypeId) -> Self {
        Self { model_id }
    }
}

impl TypeGen for GroupByResult {
    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId> {
        let model_id = self.model_id;
        t::array(
            t::struct_extends(model_id)?
                .prop(
                    "_count",
                    context.generate(&CountOutput::new(self.model_id))?, // TODO integer filter
                )
                .prop(
                    "_avg",
                    context.generate(&SelectNumbers::new(model_id, true))?,
                )
                .prop(
                    "_sum",
                    context.generate(&SelectNumbers::new(model_id, false))?,
                )
                .prop(
                    "_min",
                    context.generate(&SelectNumbers::new(model_id, false))?,
                )
                .prop(
                    "_max",
                    context.generate(&SelectNumbers::new(model_id, false))?,
                )
                .build()?,
        )
        .named(self.name())
        .build()
    }

    fn name(&self) -> String {
        let model_name = self.model_id.type_name().unwrap().unwrap();
        format!("_{}_GroupByResult", model_name)
    }
}

struct SelectNumbers {
    model_id: TypeId,
    promote_to_float: bool,
}

impl SelectNumbers {
    pub fn new(model_id: TypeId, promote_to_float: bool) -> Self {
        Self {
            model_id,
            promote_to_float,
        }
    }
}

impl TypeGen for SelectNumbers {
    fn generate(&self, _context: &mut TypeGenContext) -> Result<TypeId> {
        let mut builder = t::struct_();
        let opt_float = t::optional(t::float().build()?).build()?;
        let for_int = if self.promote_to_float {
            opt_float
        } else {
            t::optional(t::integer().build()?).build()?
        };
        for (k, type_id) in self.model_id.as_struct()?.iter_props() {
            let type_id = type_id.non_optional_concrete_type()?;
            match type_id.as_type()? {
                Type::Integer(_) => {
                    builder.prop(k, for_int);
                }
                Type::Float(_) => {
                    builder.prop(k, opt_float);
                }
                _ => {}
            }
        }

        builder.named(self.name()).build()
    }

    fn name(&self) -> String {
        let model_name = self.model_id.type_name().unwrap().unwrap();
        let suffix = if self.promote_to_float { "_1" } else { "" };
        format!("_{model_name}_SelectNumbers_{suffix}")
    }
}
