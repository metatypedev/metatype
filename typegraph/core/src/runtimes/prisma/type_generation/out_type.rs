// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::t::{self, ConcreteTypeBuilder, TypeBuilder};
use crate::types::{Type, TypeId};

use super::{Cardinality, TypeGen};

pub struct OutType {
    model_id: TypeId,
    skip_rel: Vec<String>, // list of relationships to skip to avoid infinite recursion
}

impl OutType {
    pub fn new(model_id: TypeId) -> Self {
        Self {
            model_id,
            skip_rel: vec![],
        }
    }
}

impl TypeGen for OutType {
    fn generate(
        &self,
        context: &mut super::TypeGenContext,
    ) -> crate::errors::Result<crate::types::TypeId> {
        let mut builder = t::struct_();

        for (key, type_id) in self.model_id.as_struct()?.iter_props() {
            let rel = context.registry.find_relationship_on(self.model_id, key);

            if let Some(rel) = rel {
                if self.skip_rel.contains(&rel.name) || rel.left.model_type == rel.right.model_type
                {
                    continue;
                }
                let entry = rel.get(rel.side_of_type(type_id).unwrap());
                let mut skip_rel = self.skip_rel.clone();
                skip_rel.push(rel.name.clone());

                let out_type = context.generate(&OutType {
                    model_id: entry.model_type,
                    skip_rel,
                })?;

                let out_type = match entry.cardinality {
                    Cardinality::Optional => t::optional(out_type).build()?,
                    Cardinality::One => out_type,
                    Cardinality::Many => t::array(out_type).build()?,
                };

                builder.prop(key, out_type);
            } else {
                match type_id.attrs()?.concrete_type.as_type()? {
                    Type::Func(_) => {
                        // skip, other runtime
                        continue;
                    }
                    _ => {
                        builder.prop(key, type_id);
                    }
                }
            }
        }

        builder.named(self.name()).build()
    }

    fn name(&self) -> String {
        let model_name = self.model_id.type_name().unwrap().unwrap();
        let suffix = if self.skip_rel.is_empty() {
            String::new()
        } else {
            format!("_excluding_{}", self.skip_rel.join("_"))
        };
        format!("_{model_name}OutputType{suffix}")
    }
}
