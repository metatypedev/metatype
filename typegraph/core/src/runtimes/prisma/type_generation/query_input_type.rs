// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::{
    errors::Result,
    t::{self, ConcreteTypeBuilder, TypeBuilder},
    types::TypeId,
};

use super::{
    additional_filters::{Distinct, Skip, Take},
    group_by::{GroupingFields, Having},
    order_by::OrderBy,
    query_where_expr::QueryWhereExpr,
    TypeGen, TypeGenContext,
};

pub struct QueryInputType {
    model_id: TypeId,
    is_group_by: bool,
}

impl QueryInputType {
    pub fn new(model_id: TypeId, is_group_by: bool) -> Self {
        Self {
            model_id,
            is_group_by,
        }
    }
}

impl TypeGen for QueryInputType {
    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId> {
        let mut builder = t::struct_();
        let order_by = {
            let order_by = OrderBy::new(self.model_id);
            if self.is_group_by {
                order_by.with_aggregates()
            } else {
                order_by
            }
        };

        builder
            .prop(
                "where",
                t::optional(context.generate(&QueryWhereExpr::new(self.model_id))?).build()?,
            )
            .prop(
                "orderBy",
                t::optional(context.generate(&order_by)?).build()?,
            )
            .prop("take", t::optional(context.generate(&Take)?).build()?)
            .prop("skip", t::optional(context.generate(&Skip)?).build()?);

        if self.is_group_by {
            builder.prop("by", context.generate(&GroupingFields::new(self.model_id))?);
            builder.prop(
                "having",
                // t::optional(context.generate(&Having::new(self.model_id))?).build()?,
                t::optional(context.generate(&Having::new(self.model_id))?).build()?,
            );
        } else {
            builder.prop(
                "distinct",
                t::optional(context.generate(&Distinct(self.model_id))?).build()?,
            );
        }

        builder.named(self.name()).build()
    }

    fn name(&self) -> String {
        let model_name = self.model_id.type_name().unwrap().unwrap();
        let suffix = if self.is_group_by { "_group_by" } else { "" };
        format!("_{model_name}_QueryInput{suffix}")
    }
}
