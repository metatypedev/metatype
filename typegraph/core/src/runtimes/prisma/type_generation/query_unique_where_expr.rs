// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::errors::Result;
use crate::t::{self, ConcreteTypeBuilder, TypeBuilder};
use crate::types::TypeFun;
use crate::{global_store::with_store, types::TypeId};

use super::{TypeGen, TypeGenContext};

pub struct QueryUniqueWhereExpr {
    model_id: TypeId,
}

impl QueryUniqueWhereExpr {
    pub fn new(model_id: TypeId) -> Self {
        Self { model_id }
    }
}

impl TypeGen for QueryUniqueWhereExpr {
    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId> {
        let props = with_store(|s| -> Result<_> {
            let model = s.type_as_struct(self.model_id).unwrap();

            let mut props = vec![];
            for (k, ty) in model.data.props.iter() {
                let attrs = s.get_attributes(ty.into())?;
                let is_id = attrs
                    .concrete_type
                    .as_type(s)?
                    .get_base()
                    .ok_or_else(|| "expected a concrete type".to_string())?
                    .as_id;
                let is_unique = s.get_type(ty.into())?.get_base().map_or(false, |base| {
                    base.runtime_config
                        .iter()
                        .flatten()
                        .find_map(|(k, v)| (k == "unique").then(|| v.clone()))
                        .map_or(false, |v| v == "true")
                });
                if s.is_func(attrs.concrete_type)? || (!is_id && !is_unique) {
                    continue;
                }
                props.push((k.clone(), s.resolve_quant(attrs.concrete_type)?));
            }

            Ok(props)
        })?;

        let mut st = t::struct_();
        for (k, ty) in props.into_iter() {
            st.prop(k, t::optional(ty).build()?);
        }
        st.named(self.name(context)).build()
    }

    fn name(&self, _context: &TypeGenContext) -> String {
        let name = with_store(|s| s.get_type_name(self.model_id).unwrap().unwrap().to_string());
        format!("QueryUnique{}WhereInput", name)
    }
}
