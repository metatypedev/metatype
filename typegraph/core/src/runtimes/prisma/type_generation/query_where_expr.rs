// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::global_store::with_store;
use crate::t::{self, ConcreteTypeBuilder, TypeBuilder};
use crate::{errors::Result, types::TypeId};

use super::{where_::Where, with_filters::WithFilters, TypeGen};

pub struct QueryWhereExpr {
    model_id: TypeId,
}

impl QueryWhereExpr {
    pub fn new(model_id: TypeId) -> Self {
        Self { model_id }
    }
}

impl TypeGen for QueryWhereExpr {
    fn generate(&self, context: &mut super::TypeGenContext) -> Result<TypeId> {
        let where_type = context.generate(&Where::new(self.model_id, true))?;
        let extended_type =
            context.generate(&WithFilters::new(where_type, self.model_id, false))?;

        let props = with_store(|s| extended_type.as_struct(s).unwrap().data.props.to_vec());

        let name = self.name(context);
        let self_ref = t::proxy(&name).build()?;
        let and = t::optional(t::array(self_ref).build()?).build()?;

        let mut builder = t::struct_();
        builder.named(name);
        for (k, ty) in props.into_iter() {
            builder.prop(k.clone(), ty.into());
        }
        builder.prop("AND", and);
        builder.prop("OR", and);
        builder.prop("NOT", t::optional(self_ref).build()?);

        builder.build()
    }

    fn name(&self, context: &super::TypeGenContext) -> String {
        format!(
            "Query{}WhereInput",
            context.registry.models.get(&self.model_id).unwrap().name
        )
    }
}
