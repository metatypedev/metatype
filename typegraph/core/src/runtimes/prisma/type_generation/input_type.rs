// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::errors::Result;
use crate::runtimes::prisma::type_utils::RuntimeConfig;
use crate::runtimes::prisma::{relationship::Cardinality, type_generation::where_::Where};
use crate::t::{self, ConcreteTypeBuilder, TypeBuilder};
use crate::types::{Type, TypeId};

use super::{TypeGen, TypeGenContext};

#[derive(Clone, Copy)]
enum Operation {
    Create,
    Update,
}

impl Operation {
    fn is_update(&self) -> bool {
        match self {
            Operation::Create => false,
            Operation::Update => true,
        }
    }
}

pub struct InputType {
    model_id: TypeId,
    skip_rel: Vec<String>,
    operation: Operation,
}

impl InputType {
    pub fn for_create(model_id: TypeId) -> Self {
        Self {
            model_id,
            skip_rel: vec![],
            operation: Operation::Create,
        }
    }

    pub fn for_update(model_id: TypeId) -> Self {
        Self {
            model_id,
            skip_rel: vec![],
            operation: Operation::Update,
        }
    }
}

impl TypeGen for InputType {
    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId> {
        let mut builder = t::struct_();

        for (k, type_id) in self.model_id.as_struct()?.iter_props() {
            let rel = context.registry.find_relationship_on(self.model_id, k);
            if let Some(rel) = rel {
                if self.skip_rel.contains(&rel.name) {
                    continue;
                }

                let entry = rel.get(rel.side_of_type(type_id).unwrap());
                // model
                let create = context.generate(&InputType {
                    model_id: entry.model_type,
                    skip_rel: {
                        let mut skip_rel = self.skip_rel.clone();
                        skip_rel.push(rel.name.clone());
                        skip_rel
                    },
                    operation: Operation::Create,
                })?;
                // TODO unique where
                let connect = context.generate(&Where::new(entry.model_type, false))?;

                let mut inner = t::struct_();
                inner.propx("create", t::optional(create))?;
                inner.propx("connect", t::optional(connect))?;

                if let Cardinality::Many = entry.cardinality {
                    inner.propx(
                        "createMany",
                        t::optionalx(t::struct_().propx("data", t::array(create))?)?,
                    )?;
                }

                // TODO what if cardinality is Cardinality::One ??
                builder.propx(k, t::optionalx(inner.min(1).max(1))?)?;
            } else {
                let attrs = type_id.attrs()?;
                match attrs.concrete_type.as_type()? {
                    Type::Func(_) => continue,
                    typ => {
                        let is_auto = || -> Result<_> {
                            Ok(RuntimeConfig::try_from(&typ)?.get("auto")?.unwrap_or(false))
                        };
                        builder.prop(
                            k,
                            if self.operation.is_update() || is_auto()? {
                                t::optional(type_id).build()?
                            } else {
                                type_id
                            },
                        );
                    }
                }
            }
        }

        builder.named(self.name()).build()
    }

    fn name(&self) -> String {
        let model_name = self.model_id.type_name().unwrap().unwrap();
        let suffix = if self.skip_rel.is_empty() {
            "".to_string()
        } else {
            format!("_excluding_{}", self.skip_rel.join("And"))
        };
        let op = match self.operation {
            Operation::Create => "Create",
            Operation::Update => "Update",
        };
        format!("_{model_name}_{op}Input{suffix}")
    }
}
