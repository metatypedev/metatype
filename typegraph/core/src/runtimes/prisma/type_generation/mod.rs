// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

//! This module is responsible for generating the types for prisma operations.
//!
//! Implementing the `TypeGen` trait will allow to generate and cache a type
//! in the `TypeGenContext`.
//! Type generation should always be done through the `TypeGenContext` to enable
//! the cache. Do not call `TypeGen::generate` directly.

use std::collections::HashMap;

use regex::Regex;

use self::aggregate::{CountOutput, NumberAggregateOutput};
use self::group_by::GroupByResult;
use self::input_type::InputType;
use self::out_type::OutType;
use self::query_input_type::QueryInputType;
use self::query_unique_where_expr::QueryUniqueWhereExpr;
use self::query_where_expr::QueryWhereExpr;
use self::update_input::UpdateInput;
use self::with_nested_count::WithNestedCount;

use super::relationship::registry::RelationshipRegistry;
use crate::errors::Result;
use crate::runtimes::prisma::relationship::Cardinality;
use crate::t::{self, TypeBuilder};
use crate::types::{TypeFun, TypeId};

mod additional_filters;
mod aggregate;
mod count;
pub mod group_by;
mod input_type;
mod order_by;
mod out_type;
pub mod query_input_type;
mod query_unique_where_expr;
mod query_where_expr;
mod update_input;
mod where_;
mod with_filters;
mod with_nested_count;

#[derive(Default, Debug)]
pub struct TypeGenContext {
    pub registry: RelationshipRegistry,
    cache: HashMap<String, TypeId>,
}

trait TypeGen {
    fn generate(&self, context: &mut TypeGenContext) -> Result<TypeId>;
    fn name(&self) -> String;
}

pub struct OperationTypes {
    pub input: TypeId,
    pub output: TypeId,
}

pub struct FormattedQuery {
    pub query: String,
    pub ordered_keys: Vec<String>,
}

pub fn replace_variables_to_indices(query: String, input_id: TypeId) -> Result<FormattedQuery> {
    let parameters = input_id.as_struct()?;
    let mut proc_query = query;
    let mut ordered_keys = vec![];
    let mut not_present = vec![];
    for (index, (var, _)) in parameters.data.props.iter().enumerate() {
        // Note: pattern matches ${var}, ${  var}, ${ var  }, ..
        let pattern = format!("\\$\\{{\\s*{var}\\s*\\}}");
        let re = Regex::new(&pattern).map_err(|e| e.to_string())?;
        if !re.is_match(&proc_query) {
            not_present.push(format!("{:?}", var));
        } else {
            ordered_keys.push(var.clone());
            proc_query = re
                .replace_all(&proc_query, &format!("$${}", index + 1))
                .to_string();
        }
        if !not_present.is_empty() {
            return Err(format!(
                "{} present in type definition but not in the query",
                not_present.join(", ")
            ));
        }
    }

    Ok(FormattedQuery {
        query: proc_query,
        ordered_keys,
    })
}

impl TypeGenContext {
    pub fn find_unique(&mut self, model_id: TypeId) -> Result<OperationTypes> {
        self.registry.manage(model_id)?;

        Ok(OperationTypes {
            input: t::struct_()
                .prop(
                    "where",
                    t::optional(self.generate(&QueryUniqueWhereExpr::new(model_id))?).build()?,
                )
                .build()?,
            output: t::optional(self.generate(&WithNestedCount::new(model_id))?).build()?,
        })
    }

    pub fn find_many(&mut self, model_id: TypeId) -> Result<OperationTypes> {
        self.registry.manage(model_id)?;

        Ok(OperationTypes {
            input: self.generate(&QueryInputType::new(model_id, false))?,
            output: t::array(self.generate(&WithNestedCount::new(model_id))?).build()?,
        })
    }

    pub fn find_first(&mut self, model_id: TypeId) -> Result<OperationTypes> {
        self.registry.manage(model_id)?;

        Ok(OperationTypes {
            input: self.generate(&QueryInputType::new(model_id, false))?,
            output: t::optional(self.generate(&OutType::new(model_id))?).build()?,
        })
    }

    pub fn aggregate(&mut self, model_id: TypeId) -> Result<OperationTypes> {
        self.registry.manage(model_id)?;

        Ok(OperationTypes {
            input: self.generate(&QueryInputType::new(model_id, false))?,
            // TODO typegen
            output: t::struct_()
                .prop("_count", self.generate(&CountOutput::new(model_id))?)
                .prop(
                    "_avg",
                    self.generate(&NumberAggregateOutput::new(model_id, true))?,
                )
                .prop(
                    "_sum",
                    self.generate(&NumberAggregateOutput::new(model_id, false))?,
                )
                .prop(
                    "_min",
                    self.generate(&NumberAggregateOutput::new(model_id, false))?,
                )
                .prop(
                    "_max",
                    self.generate(&NumberAggregateOutput::new(model_id, false))?,
                )
                .min(1)
                .build()?,
        })
    }

    pub fn count(&mut self, model_id: TypeId) -> Result<OperationTypes> {
        self.registry.manage(model_id)?;

        Ok(OperationTypes {
            input: self.generate(&QueryInputType::new(model_id, false))?,
            output: self.generate(&CountOutput::new(model_id))?,
        })
    }

    pub fn group_by(&mut self, model_id: TypeId) -> Result<OperationTypes> {
        self.registry.manage(model_id)?;

        Ok(OperationTypes {
            input: self.generate(&QueryInputType::new(model_id, true))?,
            output: self.generate(&GroupByResult::new(model_id))?,
        })
    }

    pub fn create_one(&mut self, model_id: TypeId) -> Result<OperationTypes> {
        self.registry.manage(model_id)?;

        Ok(OperationTypes {
            input: t::struct_()
                .prop("data", self.generate(&InputType::for_create(model_id))?)
                .build()?,
            output: self.generate(&OutType::new(model_id))?,
        })
    }

    pub fn create_many(&mut self, model_id: TypeId) -> Result<OperationTypes> {
        self.registry.manage(model_id)?;

        Ok(OperationTypes {
            input: t::struct_()
                .prop(
                    "data",
                    t::array(self.generate(&InputType::for_create(model_id))?).build()?,
                )
                .build()?,
            // TODO typegen: BatchOutput
            output: t::struct_().prop("count", t::integer().build()?).build()?,
        })
    }

    pub fn update_one(&mut self, model_id: TypeId) -> Result<OperationTypes> {
        self.registry.manage(model_id)?;

        Ok(OperationTypes {
            input: t::struct_()
                .prop("data", self.generate(&InputType::for_update(model_id))?)
                .prop(
                    "where",
                    self.generate(&QueryUniqueWhereExpr::new(model_id))?,
                )
                .build()?,
            output: self.generate(&OutType::new(model_id))?,
        })
    }

    pub fn update_many(&mut self, model_id: TypeId) -> Result<OperationTypes> {
        self.registry.manage(model_id)?;

        Ok(OperationTypes {
            input: t::struct_()
                .prop("data", self.generate(&UpdateInput::new(model_id))?)
                .prop(
                    "where",
                    t::optional(self.generate(&QueryWhereExpr::new(model_id))?).build()?,
                )
                .build()?,
            output: t::struct_().prop("count", t::integer().build()?).build()?,
        })
    }

    pub fn upsert_one(&mut self, model_id: TypeId) -> Result<OperationTypes> {
        self.registry.manage(model_id)?;

        Ok(OperationTypes {
            input: t::struct_()
                .prop(
                    "where",
                    self.generate(&QueryUniqueWhereExpr::new(model_id))?,
                )
                .prop("create", self.generate(&InputType::for_create(model_id))?)
                .prop("update", self.generate(&UpdateInput::new(model_id))?)
                // .prop("update", self.generate(&InputType::for_update(model_id))?)
                .build()?,
            output: self.generate(&OutType::new(model_id))?,
        })
    }

    pub fn delete_one(&mut self, model_id: TypeId) -> Result<OperationTypes> {
        self.registry.manage(model_id)?;
        Ok(OperationTypes {
            input: t::struct_()
                .prop(
                    "where",
                    self.generate(&QueryUniqueWhereExpr::new(model_id))?,
                )
                .build()?,
            output: self.generate(&OutType::new(model_id))?,
        })
    }

    pub fn delete_many(&mut self, model_id: TypeId) -> Result<OperationTypes> {
        self.registry.manage(model_id)?;
        Ok(OperationTypes {
            input: t::struct_()
                .prop(
                    "where",
                    t::optional(self.generate(&QueryWhereExpr::new(model_id))?).build()?,
                )
                .build()?,
            output: t::struct_().prop("count", t::integer().build()?).build()?,
        })
    }

    pub fn execute_raw(&mut self, param: TypeId) -> Result<OperationTypes> {
        let param = param.as_struct()?;
        Ok(OperationTypes {
            input: param.get_id(),
            output: t::integer().build()?,
        })
    }

    pub fn query_raw(&mut self, param: Option<TypeId>, out: TypeId) -> Result<OperationTypes> {
        let param = param.unwrap_or(t::struct_().build()?).as_struct()?;
        Ok(OperationTypes {
            input: param.get_id(),
            output: out,
        })
    }

    fn generate(&mut self, generator: &impl TypeGen) -> Result<TypeId> {
        //! Generates a type and caches it, or returns the cached type if it
        //! already exists.

        let type_name = generator.name();
        if let Some(type_id) = self.cache.get(&type_name) {
            Ok(*type_id)
        } else {
            let type_id = generator.generate(self)?;
            let typ = type_id.as_type()?;
            let name = typ
                .get_base()
                .ok_or_else(|| "Generated type must be a concrete type".to_string())?
                .name
                .as_ref()
                .ok_or_else(|| format!("Generated type must have name: {type_name}"))?;
            if name != &type_name {
                return Err(format!(
                    "Generated type name mismatch: expected {}, got {}",
                    type_name, name
                ));
            }
            self.cache.insert(type_name, type_id);
            Ok(type_id)
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::test_utils::*;
    use paste::paste;

    macro_rules! test_op {
        ( $op_name:ident, $test_inp:expr, $test_out:expr ) => {
            paste! {
                #[test]
                fn [<test _ $op_name>]() -> Result<()> {
                    test_op_body!($op_name, $test_inp, $test_out)
                }
            }
        };

        ( $op_name:ident ) => {
            test_op!($op_name, true, true);
        };

        ( $op_name:ident, input_only ) => {
            test_op!($op_name, true, false);
        };

        ( $op_name:ident, output_only ) => {
            test_op!($op_name, false, true);
        };
    }

    macro_rules! test_op_body {
        ( $op_name:ident, $test_inp:expr, $test_out:expr ) => {{
            let mut context = TypeGenContext::default();

            let record = models::simple_record()?;
            context.registry.manage(record)?;
            let types = context.$op_name(record)?;
            if $test_inp {
                insta::assert_snapshot!(
                    concat!(stringify!($op_name), " Record inp"),
                    tree::print(types.input)
                );
            }
            if $test_out {
                insta::assert_snapshot!(
                    concat!(stringify!($op_name), " Record out"),
                    tree::print(types.output)
                );
            }

            let (user, post) = models::simple_relationship()?;
            context.registry.manage(user)?;

            let types = context.$op_name(user)?;
            if $test_inp {
                insta::assert_snapshot!(
                    concat!(stringify!($op_name), " User inp"),
                    tree::print(types.input)
                );
            }

            if $test_out {
                insta::assert_snapshot!(
                    concat!(stringify!($op_name), " User out"),
                    tree::print(types.output)
                );
            }

            let types = context.$op_name(post)?;
            if $test_inp {
                insta::assert_snapshot!(
                    concat!(stringify!($op_name), " Post inp"),
                    tree::print(types.input)
                );
            }
            if $test_out {
                insta::assert_snapshot!(
                    concat!(stringify!($op_name), " Post out"),
                    tree::print(types.output)
                );
            }

            Ok(())
        }};
    }

    test_op!(find_unique);
    test_op!(find_many);
    test_op!(find_first, output_only);
    test_op!(aggregate, output_only);
    test_op!(group_by);
    test_op!(create_one, input_only);
    test_op!(create_many);
    test_op!(update_one, input_only);
    test_op!(update_many, input_only);
    // the following operations reuse already tests types, so no need to test them
    // test_op!(count);
    // test_op!(upsert_one);
    // test_op!(delete_one);
    // test_op!(delete_many);
}
