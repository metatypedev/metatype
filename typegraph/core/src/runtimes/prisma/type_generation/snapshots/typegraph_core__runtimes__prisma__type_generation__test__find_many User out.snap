---
source: typegraph/core/src/runtimes/prisma/type_generation/mod.rs
expression: tp.print(out)
---
root: list '_126_UserWithNestedCount[]' #126
└─ item: struct 'UserWithNestedCount' #125
   ├─ [_count]: struct #124
   │  └─ [posts]: optional '_Count' #123
   │     └─ item: integer #122
   ├─ [id]: integer #74
   ├─ [name]: string #75
   └─ [posts]: list '_121_PostWithNestedCount_excluding___rel_Post_User_1[]' #121
      └─ item: struct 'PostWithNestedCount_excluding___rel_Post_User_1' #120
         ├─ [id]: integer #79
         └─ [title]: string #80

